chain/network/src/peer.rs

```diff
@@ -597,22 +607,30 @@ impl StreamHandler<Vec<u8>> for Peer {
 
                 if handshake.version != PROTOCOL_VERSION {
                     info!(target: "network", "Received connection from node with different network protocol version.");
-                    ctx.address().do_send(SendMessage {
-                        message: PeerMessage::HandshakeFailure(
-                            self.node_info.clone(),
-                            HandshakeFailureReason::ProtocolVersionMismatch(PROTOCOL_VERSION),
-                        ),
-                    });
+                    self.send_message(PeerMessage::HandshakeFailure(
+                        self.node_info.clone(),
+                        HandshakeFailureReason::ProtocolVersionMismatch(PROTOCOL_VERSION),
+                    ));
                     return;
                     // Connection will be closed by a handshake timeout
                 }
 
                 if handshake.peer_id == self.node_info.id {
-                    warn!(target: "network", "Received info about itself. Disconnecting this peer.");
+                    debug!(target: "network", "Received info about itself. Disconnecting this peer.");
                     ctx.stop();
                     return;
                 }
 
+                if handshake.target_peer_id != self.node_info.id {
+                    debug!(target: "network", "Received handshake from {:?} to {:?} but I am {:?}", handshake.peer_id, handshake.target_peer_id, self.node_info.id);
```

Comment:
This is the central fix of this PR and why it can happens is in the description of the PR. In practice it happens, because you start your node with one `node_key` (A), then you remove `~/.near` folder and recreate it with different `node_key` (B) and connect to the network. All nodes (even you) still believe that A exists at that address and will try to establish connection, but instead of A, they will find B there.